# SQL Basics for Data Science: Project (15 Nov 2020)

## Background Information


You are a Data Science Consultant working on a project to understand car crashes in the city of Chicago. You are required to make recommendations to the office of the mayor of Chicago on the steps to take that would lead to a reduction in car crashes, based on the findings from the dataset analysis.

### Problem Statement

As part of your descriptive analysis report, you decide to find answers to the following questions using SQL.
1. Display a list of 5 makes that had the highest number of crashes?
2. Most personal vehicles involved in crashes were travelling towards which direction? 
3. How many passengers were involved in personal cars were in accidents?
4. How many passengers were involved in accidents while travelling in personal Sport Utility vehicles?
5. Compute the total number of Mercedez-Benz vehicles involved in accidents while being parked.
6. Compute the total occupant count per vehicle count.
7. Sort the number of passengers in descending order by travel direction taking into account travel direction. Return travel direction and no. of passengers.
8. Select top 5 vehicle models involved in the accident with no defect, order them by make and show their manoeuvre. 
9. Which ridesharing vehicles in an accident had the most number of passengers? Return model, and no of passengers.
 
<strong>Remember:</strong> What matters the most is what your insights mean to the business. Make sure to have a section where you elaborate how your analysis findings tie to your recommendations.


## Dataset Explained


Before performing any analysis, you should select only vehicles manufactured in the last 15 years from MySQL traffic_accidents_database whose URL and credentials are given below.

URL = http://************/phpmyadmin/<br>
username = learner <br>
password = `******************`



## Data Preparation

After connecting to the database using the credentials provided, I have selected the required set of records by executing the following SQL query.

SELECT * FROM `traffic_accidents` WHERE vehicle_year > 2005  AND vehicle_year <= 2020

I have then exported the resulting recordset as a .csv file `traffic_accidents.csv` for downstream analysis using SQL.

## Step 1. Importing Required Libraries

In [12]:
# Importing pandas
import pandas as pd

# Loading SQL extension
%load_ext sql

# Connecting sqlite database
%sql sqlite://

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


'Connected: @None'

## Step 2. Loading Dataset

In [13]:
# Load the dataset from a csv file and store it in a dataframe called accidents.
accidents = pd.read_csv('traffic_accidents.csv')


# Store the dataset in the in-memory sqlite database. 
# for controls, check first if the table exists in the database, and if so drop it. 
%sql DROP TABLE if EXISTS accidents;


# create an SQL table in the in-memory sqlite database and store 
# the recordset in readiness for further analysis with SQL
%sql PERSIST accidents; 

 * sqlite://
Done.
 * sqlite://


'Persisted accidents'

### 2.1 Sample Recordset

In [14]:
# Sample the first few records of the created dataset to understand its structure.

%%sql 
SELECT * FROM accidents LIMIT 5

 * sqlite://
Done.


[(0, 758369, 'JC469557', '2019-10-12T01:57:00', 1, 'DRIVER', None, 720366, None, 'DODGE', 'NITRO', 'IL', 2011, 'UNKNOWN', 'PASSENGER', 'UNKNOWN/NA', 'W', 'STRAIGHT AHEAD', None, None, 1, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 1, None, 'OTHER', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None),
 (1, 758356, 'JC469523', '2019-10-12T00:30:00', 1, 'DRIVER', None, 720356, None, 'VOLVO', 'XC90', 'IL', 2020, 'NONE', 'SPORT UTILITY VEHICLE (SUV)', 'PERSONAL', 'S', 'SLOW/STOP IN TRAFFIC', None, None, 1, None, None, None, None, None, None, 1, None, None, None, None, None, None, None, None, None, None, 'SIDE-RIGHT', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None),
 (2, 758329, 'JC469467', '2019-10-11T23:41:00', 1, 'DRIVER', 2.0, 720325, None, 'TOYOTA MOTOR COMPANY, LTD.', 'UNKNOWN', 'IL', 2015, 'NONE', 'PASSENGER', 'PERSONAL', 'E', 'PASSING/OVERTAKING', 1, None, 3, None, 'unknown', 'unknown', None, None, None, 1, 1, 1, 1, 1, None, None, None, None, None, None, 'REAR', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None),
 (3, 758330, 'JC469467', '2019-10-11T23:41:00', 2, 'DRIVER', None, 720334, None, 'SUBARU', 'FORESTER', 'IL', 2014, 'NONE', 'PASSENGER', 'PERSONAL', 'E', 'TURNING LEFT', None, None, 1, None, None, None, None, 1, 1, None, None, None, None, None, None, None, None, None, 1, None, 'FRONT', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None),
 (4, 758333, 'JC469482', '2019-10-11T23:20:00', 1, 'DRIVER', None, 720328, None, 'CHEVROLET', 'IMPALA', 'IL', 2017, 'UNKNOWN', 'PASSENGER', 'PERSONAL', 'N', 'STRAIGHT AHEAD', None, None, 1, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'UNKNOWN', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None)]

## Step 3. Business Needs

**1. Display a list of 5 makes that had the highest number of crashes?**

In [15]:
# we make use of the property 'make' and use the COUNT()function against this 
# variable for each of its make group. 

%%sql 
SELECT make as 'Car Make', COUNT(make) as 'Accident Frequency'
FROM accidents
GROUP BY make
ORDER BY COUNT(make) desc
LIMIT 5;

 * sqlite://
Done.


[('TOYOTA MOTOR COMPANY, LTD.', 609),
 ('FORD', 545),
 ('CHEVROLET', 538),
 ('NISSAN', 478),
 ('HONDA', 345)]

'TOYOTA MOTOR COMPANY, LTD' with 609 accidents had the highest number of accidents reported for the period betweeen 2005 and 2020, followed by FORD with 545, CHEVROLET with 538, NISSAN with 478 and HONDA at 5th position with 345 accidents reported.



---



**2. Towards which direction were the most personal vehicles involved in crashes travelling?**

In [16]:
# we use 'vehicle_type' and 'travel_direction' to pick personal vehicles and tdirection accordingly. 
# further we use COUNT() funtion for each of the directions to determine the mode of direction.
%%sql
SELECT travel_direction as 'Travel Direction', COUNT(travel_direction) as 'Crash Frequency'
FROM accidents
WHERE vehicle_use = 'PERSONAL'
GROUP BY travel_direction 
ORDER BY COUNT(travel_direction) desc
LIMIT 5;

 * sqlite://
Done.


[('N', 849), ('S', 807), ('W', 765), ('E', 708), ('UNKNOWN', 116)]

The highest number of passenger cars, at 849, reported with accidents were travelling towards the North.



---



**3. How many passengers were involved in personal cars accidents??**

In [17]:
# we use the SUM() function to get the total number of passengers for all the 
# personal cars involved in accidents.
%%sql
SELECT SUM(CAST(num_passengers AS integer)) as 'No of Passengers'
FROM accidents
WHERE vehicle_use = 'PERSONAL'

 * sqlite://
Done.


[(827,)]

A total of 827 passengers were involved in personal car accidents.



---



**4. How many passengers were involved in accidents while travelling in personal Sport Utility vehicles?**

In [20]:
# Use the SUM() function to get the total number of passengers for all the 
# personal sport utility vehicles involved in accidents.
%%sql
SELECT SUM(CAST(num_passengers AS interger)) as 'No of Passengers'
FROM accidents
WHERE vehicle_use = 'PERSONAL' AND vehicle_type = 'SPORT UTILITY VEHICLE (SUV)'


 * sqlite://
Done.


[(137,)]

A total of 137 passengers were involved in personal SUV car accidents. This accounts for about 16.5% of all the personal cars accidents.



---



**5. Compute the total number of Mercedez-Benz vehicles involved in accidents while being parked.**

In [21]:
#Use COUNT() function to get the total number of Mercedez-Benz cars.

%%sql
SELECT COUNT(make) as 'No of Mercedez-Benz'
FROM accidents
WHERE make = 'MERCEDES-BENZ' AND maneuver = 'PARKED'

 * sqlite://
Done.


[(10,)]

Only 10 Mercedez-Benz cars were involved in an accident while parked.



---



**6.Compute the total occupancy count per vehicle count.**

In [22]:
# I have interpreted "per vehicle count" to mean for each make of the vehicle.
# Use the SUM() function to get the total number of passengers for each of the makes of the vehicles.
%%sql
SELECT make, SUM(occupant_cnt) as 'Occupancy'
FROM accidents
GROUP BY make
ORDER BY SUM(occupant_cnt) DESC

 * sqlite://
Done.


[('TOYOTA MOTOR COMPANY, LTD.', 669),
 ('FORD', 627),
 ('CHEVROLET', 590),
 ('NISSAN', 553),
 ('HONDA', 364),
 ('HYUNDAI', 303),
 ('DODGE', 280),
 ('JEEP', 255),
 ('KIA MOTORS CORP', 176),
 ('CHRYSLER', 135),
 ('VOLKSWAGEN', 109),
 ('UNKNOWN', 92),
 ('BUICK', 88),
 ('MAZDA', 81),
 ('GENERAL MOTORS CORP.', 74),
 ('LEXUS', 73),
 ('SUBARU', 70),
 ('MERCEDES-BENZ', 64),
 ('AUDI', 60),
 ('NEW FLYER', 57),
 ('INFINITI', 56),
 ('BMW', 55),
 ('LINCOLN-CONTINENTAL', 53),
 ('CADILLAC', 49),
 ('FREIGHTLINER CORP.', 49),
 ('PONTIAC', 46),
 ('ACURA', 34),
 ('MITSUBISHI', 32),
 ('VOLVO', 29),
 ('SATURN', 19),
 ('SCION', 17),
 ('ACURA (DIV. OF AMERICAN HONDA MOTOR CO.)', 15),
 ('ICL CARGO TRAILER', 15),
 ('BLUEBIRD INTERNATIONAL', 13),
 ('MINI', 13),
 ('LAND ROVER', 12),
 ('PREVOST CAR, BUS, SAINTE CLAIRE, QUEBEC', 12),
 ('ISUZU', 11),
 ('KENWORTH MOTOR TRUCK CO.', 10),
 ('INTERNATIONAL TRAILER CORP.', 9),
 ('PORSCHE', 8),
 ('SUZUKI', 8),
 ('AMTRAN CORP, GENSIS BUSES', 7),
 ('BLUE BIRD BODY CO.', 7),
 ('MERCURY', 7),
 ('AUTOCAR', 6),
 ('MACK TRUCKS, INC.', 6),
 ('PETERBILT MOTORS CO., DIVISION PACCAR, INC.', 6),
 ('FIAT', 5),
 ('HINO', 5),
 ('INTERNATIONAL COACH MFG.', 5),
 ('JAGUAR', 5),
 ('TESLA MOTORS', 5),
 ("FLEET CAP'N TRAILERS", 4),
 ('INTERNATIONAL HARVESTER', 4),
 ('KAWASAKI MOTORS CORP.', 4),
 ('AM GENERAL/HUMMER', 3),
 ('EL DORADO INTERNATIONAL, INC.; EL DORADO MFG., LTD.', 3),
 ('FIESTA (IMPORTED BY FORD)', 3),
 ('GILL MFG. CO.', 3),
 ('HARLEY-DAVIDSON', 3),
 ('INTENATIONAL INDUSTRIES, INC.', 3),
 ('RANGE ROVER OF NORTH AMERICA', 3),
 ('SAAB', 3),
 ('MOTOR COACH INDUSTRIES, INC.', 2),
 ('WABCO CONSTRUCTION & MINING EQUIPMENT GROUP', 2),
 ('(UNLISTED MANUFACTURER)', 1),
 ('AMERICAN TRAILER & MFG. CO. (SEE ALSO TRANSPORTER)', 1),
 ('ASTON-MARTIN', 1),
 ('BANTAM CRANE', 1),
 ('CALDWELL DIV., SMITH INTERNATIONAL, INC.', 1),
 ('CASCADE COACH CO.', 1),
 ('CHEVELLE MOBILE HOMES, INC.', 1),
 ('CONTINENTAL', 1),
 ('EIDAL INTERNATIONAL CORP.', 1),
 ('FLEETCRAFT CORP.', 1),
 ('FLEETWOOD ENTERPRISES', 1),
 ('FONTAINE TRUCK EQUIPMENT CO.', 1),
 ('GOODYEAR CAMPER MFG., INC., LINCOLN, NB', 1),
 ('GREAT DANE TRAILERS INC.', 1),
 ('HAULER-TRAILER(MFD. BY NORDINE MFG.)', 1),
 ('HINDUSTAN', 1),
 ('LINCOLN ELECTRIC CO., INC. CLEVELAND, OH ; LINCOLN WELDER', 1),
 ('MACK TRAILER MFG. CO., INC., DETRIOT, MI', 1),
 ('NAVISTAR', 1),
 ('NISSAN DIESEL MOTOR CO.', 1),
 ('OLDSMOBILE', 1),
 ('ROLLS INTERNATIONAL TRAILER(MOBILE HOME)', 1),
 ('SCOOTALONG', 1),
 ('STERLINE FLATBED TRAILER', 1),
 ('STERLING', 1),
 ('THOMAS BUILT BUS CO.', 1),
 ('TOYOCAR VAN CONTAINER TRAILER', 1),
 ('WORK HORSE MFG.', 1),
 ('YAMAHA', 1),
 ('ALFA ROMEO', 0),
 ('BENTLEY', 0),
 ('FERRARI', 0),
 ('QUALITY TRAILER', 0),
 ('TOWMASTER', 0),
 ('TRIUMPH', 0)]

*I have interpreted **"per vehicle count"** to mean for each make of the vehicle*

While some vehicles, about 6 makes of cars have not had any reported accidents, Toyota and Ford topped the list of vehicles with the highest count of passengers.

 

---



**7. Sort the number of passengers in descending order by travel direction taking into account travel direction.**

In [24]:
# We will need data from  'occupant_cnt', travel_direction, and 'travel_direction' 
# to respond to the need above. 
# further we use COUNT() function on occupant_cnt to get the number of passengers 
# for each of the travel directions
%%sql
SELECT travel_direction as 'Travel Direction', COUNT(num_passengers) as 'No. of Passengers'
FROM accidents
GROUP BY travel_direction 
ORDER BY COUNT(num_passengers) desc, travel_direction

 * sqlite://
Done.


[('N', 198),
 ('W', 194),
 ('E', 184),
 ('S', 184),
 ('SE', 12),
 ('UNKNOWN', 9),
 ('NE', 6),
 ('NW', 6),
 ('SW', 6)]

Travel direction `NORTH` recorded the highest number of passengers involved in accidents with `198` people.



---



**8. Select top 5 vehicle models involved in the accident with no defect, order them by make and show their manoeuvre.**

In [27]:
# We use 'model' and 'vehicle_defect', 'make' and maneuver to respond to the need above
# but pick only cases where vehicles didnt have any defects. 
# further we use ORDER BY Clause for sorting by 'make' and finally LIMIT as we are only interested in the top 5
%%sql
SELECT make, model, maneuver, COUNT(model) as 'vehicle model'
FROM accidents
WHERE vehicle_defect = 'NONE'
GROUP BY make, model
ORDER BY COUNT(model) DESC, make ASC
LIMIT 5

 * sqlite://
Done.


[('TOYOTA MOTOR COMPANY, LTD.', 'CAMRY', 'TURNING RIGHT', 129),
 ('NISSAN', 'NISSAN ALTIMA', 'PARKED', 78),
 ('NISSAN', 'SENTRA (DATSUN AND NISSAN HAVE MERGED)', 'STRAIGHT AHEAD', 71),
 ('CHEVROLET', 'MALIBU (CHEVELLE)', 'PARKED', 68),
 ('FORD', 'UNKNOWN', 'BACKING', 60)]

From the analysis, `Toyota Camry` vehicles which had no defects registered the highest number of accidents at `129` accidents.



---



**9. Which ridesharing vehicle models involved accidents had the most number of passengers? Return model, and no of passengers?**

In [28]:
# We use 'model', 'num_passengers' and 'vehicle_use' to pick relevant records. 
# further we use COUNT() funtion for each of the directions to determine the mode of direction.
%%sql
SELECT model, COUNT(num_passengers) as 'no. of passengers'
FROM accidents
WHERE vehicle_use = 'RIDESHARE SERVICE'
GROUP BY model 
ORDER BY COUNT(num_passengers) DESC
LIMIT 5;

 * sqlite://
Done.


[('UNKNOWN', 4),
 ('CAMRY', 3),
 ('COROLLA', 3),
 ('NISSAN ALTIMA', 3),
 ('FUSION', 2)]

The ridesharing model of vehicles which had most number of passengers were, unfortunately, not classified but registered only as "Unknown". In the second position is Toyota Camry which, from previous analysis contributed to the most number of passengers involved in accidents.

## Step 4. Recommendations

From the foregoing analysis, we make the following observations:
<ul>
1. Toyota Motor Company and Ford makes of vehcles top the list of cars with accidents, having reported 609 and 545 accidents respectively. 

2. Personal cars travelling towards the North and South had the most accidents. (This seems to be basically the same roads but different directions, a pointer to probably poor conditions of the roads, other physical features along the roads with blackspots etc. )

3. Of all personal car accidents, SUV's contribute about 16.5%.

4. The ridesharing model of vehicles which had most number of passengers were, unfortunately, not classified but registered only as "Unknown". In the second position is Toyota Camry which, from previous analysis contributed to the most number of passengers involved in accidents</font></em>
</ul>
<p>
<br>

**Recommendations:** From the available data, we are neither able to deduce of infer the real causes of the accidents and so would recommend that the authorities carry out deeper root-cause analysis (RCA) on the following; 

1.   Why Toyota and Ford make of vehicles are the most frequently make of   cars involved in crashes. (Could it be because of the fact that there are more of such cars on the road than other makes hence more probable to make crashes?)

2.   Why cars travelling to the North and to the South are the most frequently cars involved in crashes. (Could it be the state of the North-South Road?, other associated road users along the road like many pedestrians or large haulier trucks? etc.) 
